src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlObjectSerializerReadContext.cs

```diff
@@ -361,9 +361,9 @@ public void ReplaceDeserializedObject(string id, object? oldObj, object? newObj)
                 // throw in such cases to allow us add fix-up support in the future if we need to.
                 if (DeserializedObjects.IsObjectReferenced(id))
                 {
-                    // https://github.com/dotnet/runtime/issues/41465 - oldObj or newObj may be null below - suppress compiler error by asserting non-null
-                    Debug.Assert(oldObj != null && newObj != null);
-                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(XmlObjectSerializer.CreateSerializationException(SR.Format(SR.FactoryObjectContainsSelfReference, DataContract.GetClrTypeFullName(oldObj.GetType()), DataContract.GetClrTypeFullName(newObj.GetType()), id)));
+                    var oldType = (oldObj != null) ? DataContract.GetClrTypeFullName(oldObj.GetType()) : "unknown type";
+                    var newType = (newObj != null) ? DataContract.GetClrTypeFullName(newObj.GetType()) : "unknown type";
```

Comment:
Nit:
```suggestion
                    string oldType = (oldObj != null) ? DataContract.GetClrTypeFullName(oldObj.GetType()) : "unknown type";
                    string newType = (newObj != null) ? DataContract.GetClrTypeFullName(newObj.GetType()) : "unknown type";
```
Also, does "unknown type" need to come from a resx?